<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>APCalign • APCalign</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="APCalign">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">APCalign</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/APCalign.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/articles/data-providers.html">Data sources (APC &amp; APNI)</a></li>
    <li><a class="dropdown-item" href="../articles/updating-taxon-names.html">Taxon matching</a></li>
    <li><a class="dropdown-item" href="../articles/reproducibility.html">Using APC versions for reproducibility</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/traitecoevo/APCalign/"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>APCalign</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/traitecoevo/APCalign/blob/2025-08-19/vignettes/APCalign.Rmd" class="external-link"><code>vignettes/APCalign.Rmd</code></a></small>
      <div class="d-none name"><code>APCalign.Rmd</code></div>
    </div>

    
    
<p>When working with biodiversity data, it is important to verify
taxonomic names with an authoritative list and correct any out-of-date
names. The ‘APCalign’ package simplifies this process by:</p>
<ul>
<li>Accessing up-to-date taxonomic information from the <a href="https://biodiversity.org.au/nsl/services/search/taxonomy" class="external-link">Australian
Plant Census</a> and the <a href="https://biodiversity.org.au/nsl/services/search/names" class="external-link">Australia
Plant Name Index</a>.</li>
<li>Aligning authoritative names to your taxonomic names using our <a href="https://traitecoevo.github.io/APCalign/articles/updating-taxon-names.html">fuzzy
matching algorithm</a>
</li>
<li>Updating your taxonomic names in a transparent, reproducible
manner</li>
</ul>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>‘APCalign’ is currently not on CRAN. You can install its current
developmental version using</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"traitecoevo/APCalign"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://traitecoevo.github.io/APCalign/">APCalign</a></span><span class="op">)</span></span></code></pre></div>
<p>To demonstrate how to use ‘APCalign’, we will use an example dataset
<code>gbif_lite</code> which is documented in
<code><a href="../reference/gbif_lite.html">?gbif_lite</a></code></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">gbif_lite</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 129   7</span></span>
<span></span>
<span><span class="va">gbif_lite</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 129 × 7</span></span>
<span><span class="co">#&gt;   species              infraspecificepithet taxonrank decimalLongitude decimalLatitude scientificname verbatimscientificname</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                &lt;chr&gt;                &lt;chr&gt;                &lt;dbl&gt;           &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                 </span></span>
<span><span class="co">#&gt; 1 Tetratheca ciliata   &lt;NA&gt;                 SPECIES               145.           -37.4 Tetratheca ci… Tetratheca ciliata    </span></span>
<span><span class="co">#&gt; 2 Peganum harmala      &lt;NA&gt;                 SPECIES               139.           -33.3 Peganum harma… Peganum harmala       </span></span>
<span><span class="co">#&gt; 3 Calotis multicaulis  &lt;NA&gt;                 SPECIES               115.           -24.3 Calotis multi… Calotis multicaulis   </span></span>
<span><span class="co">#&gt; 4 Leptospermum triner… &lt;NA&gt;                 SPECIES               151.           -34.0 Leptospermum … Leptospermum trinervi…</span></span>
<span><span class="co">#&gt; 5 Lepidosperma latera… &lt;NA&gt;                 SPECIES               142.           -37.3 Lepidosperma … Lepidosperma laterale </span></span>
<span><span class="co">#&gt; 6 Enneapogon polyphyl… &lt;NA&gt;                 SPECIES               129.           -17.8 Enneapogon po… Enneapogon polyphyllus</span></span>
<span><span class="co">#&gt; # ℹ 123 more rows</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="retrieve-taxonomic-resources">Retrieve taxonomic resources<a class="anchor" aria-label="anchor" href="#retrieve-taxonomic-resources"></a>
</h2>
<p>The first step is to retrieve the entire APC and APNI name databases
and store them locally as taxonomic resources. We achieve this using
<code><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources()</a></code>.</p>
<p>There are two versions of the databases that you can retrieve with
the <code>stable_or_current_data</code> argument. Calling:</p>
<ul>
<li>
<code>stable</code> will retrieve the most recent, archived version
of the databases from our <a href="https://github.com/traitecoevo/APCalign/releases" class="external-link">GitHub
releases</a>. This is set as the default option.</li>
<li>
<code>current</code> will retrieve the up-to-date databases directly
from the APC and APNI website.</li>
</ul>
<p>Note that the databases are quite large so the initial retrieval of
<code>stable</code> versions will take a few minutes. Once the taxonomic
resources have been stored locally, subsequent retrievals will take less
time. Retrieving <code>current</code> resources will always take longer
since it is accessing the latest information from the website. Check out
our <a href="https://traitecoevo.github.io/APCalign/articles/caching.html">Resource
Caching</a> article to learn more about how the APC and APNIC databases
are accessed, stored and retrieved.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Benchmarking the retrieval of `stable` or `current` resources</span></span>
<span><span class="va">stable_start_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">stable_resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span>stable_or_current_data <span class="op">=</span> <span class="st">"stable"</span><span class="op">)</span></span></code></pre></div>
<pre><code>#&gt; Loading resources into memory...
#&gt; 
=========================================
===================================================================================
============================================================================================================================
#&gt; ...done
stable_end_time &lt;-  Sys.time()

current_start_time &lt;- Sys.time()
current_resources &lt;- load_taxonomic_resources(stable_or_current_data = "current")</code></pre>
<pre><code>#&gt; Loading resources into memory...
#&gt; 
=========================================
===================================================================================
============================================================================================================================
#&gt; ...done
current_end_time &lt;-  Sys.time()

# Compare times
stable_end_time - stable_start_time
#&gt; Time difference of 6.69976 secs</code></pre>
<p>For a more reproducible workflow, we recommend specifying the exact
<code>stable</code> version you want to use.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span>stable_or_current_data <span class="op">=</span> <span class="st">"stable"</span>, version <span class="op">=</span> <span class="st">"2024-10-11"</span><span class="op">)</span></span></code></pre></div>
<pre><code>#&gt; Loading resources into memory...
#&gt; 
=========================================
===================================================================================
============================================================================================================================
#&gt; ...done</code></pre>
</div>
<div class="section level2">
<h2 id="align-and-update-plant-taxon-names">Align and update plant taxon names<a class="anchor" aria-label="anchor" href="#align-and-update-plant-taxon-names"></a>
</h2>
<p>Now we can query our taxonomic names against the taxonomic resources
we just retrieved using <code><a href="../reference/create_taxonomic_update_lookup.html">create_taxonomic_update_lookup()</a></code>.
This all-in-one function will:</p>
<ul>
<li>Align your taxonomic names to APC and APNI using our <a href="https://traitecoevo.github.io/APCalign/articles/updating-taxon-names.html">matching
algorithms</a><br>
</li>
<li>Update names to an APC-accepted species or infraspecific name
whenever possible.<br>
</li>
<li>Return a suggested name for all names, defaulting to an
<code>accepted_name</code> when available, and otherwise providing an
APNI name or a name where only a genus-level alignment is possible.</li>
</ul>
<p>If you would like to learn more about each of these step, take a look
at the section <a href="#closer-look">Closer look at name alignment and
updating with ‘APCalign’</a></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:testthat':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     matches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span></span>
<span><span class="va">updated_gbif_names</span> <span class="op">&lt;-</span> <span class="va">gbif_lite</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">species</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/create_taxonomic_update_lookup.html">create_taxonomic_update_lookup</a></span><span class="op">(</span>resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking alignments of 121 taxa</span></span>
<span><span class="co">#&gt;   -&gt; of these 112 names have a perfect match to a scientific name in the APC. </span></span>
<span><span class="co">#&gt;       Alignments being sought for remaining names.</span></span>
<span></span>
<span><span class="va">updated_gbif_names</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 129 × 12</span></span>
<span><span class="co">#&gt;   original_name           aligned_name      accepted_name suggested_name genus taxon_rank taxonomic_dataset taxonomic_status</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                   &lt;chr&gt;             &lt;chr&gt;         &lt;chr&gt;          &lt;chr&gt; &lt;chr&gt;      &lt;chr&gt;             &lt;chr&gt;           </span></span>
<span><span class="co">#&gt; 1 Tetratheca ciliata      Tetratheca cilia… Tetratheca c… Tetratheca ci… Tetr… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 2 Peganum harmala         Peganum harmala   Peganum harm… Peganum harma… Pega… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 3 Calotis multicaulis     Calotis multicau… Calotis mult… Calotis multi… Calo… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 4 Leptospermum trinervium Leptospermum tri… Leptospermum… Leptospermum … Lept… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 5 Lepidosperma laterale   Lepidosperma lat… Lepidosperma… Lepidosperma … Lepi… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 6 Enneapogon polyphyllus  Enneapogon polyp… Enneapogon p… Enneapogon po… Enne… species    APC               accepted        </span></span>
<span><span class="co">#&gt; # ℹ 123 more rows</span></span>
<span><span class="co">#&gt; # ℹ 4 more variables: scientific_name &lt;chr&gt;, aligned_reason &lt;chr&gt;, update_reason &lt;chr&gt;, number_of_collapsed_taxa &lt;dbl&gt;</span></span></code></pre></div>
<p>The <code>original_name</code> is the taxon name used in your
original data. The <code>aligned_name</code> is the taxon name we used
to link with the APC to identify any synonyms. The
<code>accepted_name</code> is the currently, accepted taxon name used by
the Australian Plant Census. The <code>suggested_name</code> is the best
possible name option for the <code>original_name</code>.</p>
</div>
<div class="section level2">
<h2 id="plant-established-status-across-statesterritories">Plant established status across states/territories<a class="anchor" aria-label="anchor" href="#plant-established-status-across-statesterritories"></a>
</h2>
<p>‘APCalign’ can also provide the state/territory distribution for
established status (native/introduced) from the APC.</p>
<p>We can access the established status data by state/territory using
<code><a href="../reference/create_species_state_origin_matrix.html">create_species_state_origin_matrix()</a></code></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Retrieve status data by state/territory </span></span>
<span><span class="va">status_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_species_state_origin_matrix.html">create_species_state_origin_matrix</a></span><span class="op">(</span>resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span></span></code></pre></div>
<p>Here is a breakdown of all possible values for
<code>origin</code></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'purrr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:testthat':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     is_null</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sfirke/janitor" class="external-link">janitor</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'janitor'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     chisq.test, fisher.test</span></span>
<span></span>
<span><span class="co"># Obtain unique values</span></span>
<span><span class="va">status_matrix</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">species</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">flatten_chr</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://sfirke.github.io/janitor/reference/tabyl.html" class="external-link">tabyl</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;  flatten_chr(select(status_matrix, -species))      n      percent</span></span>
<span><span class="co">#&gt;                        doubtfully naturalised   1129 2.387326e-03</span></span>
<span><span class="co">#&gt;                          formerly naturalised    277 5.857302e-04</span></span>
<span><span class="co">#&gt;                                        native  40363 8.534956e-02</span></span>
<span><span class="co">#&gt;             native and doubtfully naturalised      9 1.903094e-05</span></span>
<span><span class="co">#&gt;                        native and naturalised    137 2.896933e-04</span></span>
<span><span class="co">#&gt;                   native and uncertain origin      2 4.229099e-06</span></span>
<span><span class="co">#&gt;                                   naturalised   8769 1.854248e-02</span></span>
<span><span class="co">#&gt;                                   not present 422103 8.925576e-01</span></span>
<span><span class="co">#&gt;                              presumed extinct    102 2.156840e-04</span></span>
<span><span class="co">#&gt;                              uncertain origin     23 4.863464e-05</span></span></code></pre></div>
<!-- The formal definitions of the various established status can be found at XX.  -->
<p>You can also obtain the breakdown of species by established status
for a particular state/territory using
<code><a href="../reference/state_diversity_counts.html">state_diversity_counts()</a></code></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/state_diversity_counts.html">state_diversity_counts</a></span><span class="op">(</span><span class="st">"NSW"</span>, resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 7 × 3</span></span>
<span><span class="co">#&gt;   origin                            state num_species</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                             &lt;chr&gt; &lt;table[1d]&gt;</span></span>
<span><span class="co">#&gt; 1 doubtfully naturalised            NSW     93       </span></span>
<span><span class="co">#&gt; 2 formerly naturalised              NSW      8       </span></span>
<span><span class="co">#&gt; 3 native                            NSW   5968       </span></span>
<span><span class="co">#&gt; 4 native and doubtfully naturalised NSW      2       </span></span>
<span><span class="co">#&gt; 5 native and naturalised            NSW     34       </span></span>
<span><span class="co">#&gt; 6 naturalised                       NSW   1580       </span></span>
<span><span class="co">#&gt; 7 presumed extinct                  NSW      8</span></span></code></pre></div>
<p>Using the established status data and state/territory information, we
can check if a plant taxa is a native using
<code><a href="../reference/native_anywhere_in_australia.html">native_anywhere_in_australia()</a></code></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">updated_gbif_names</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html" class="external-link">sample_n</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="co"># Choosing a random species</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">suggested_name</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Extracting this APC accepted name</span></span>
<span>  <span class="fu"><a href="../reference/native_anywhere_in_australia.html">native_anywhere_in_australia</a></span><span class="op">(</span>resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span> </span>
<span><span class="co">#&gt; # A tibble: 1 × 2</span></span>
<span><span class="co">#&gt;   species             native_anywhere_in_aus</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;               &lt;chr&gt;                 </span></span>
<span><span class="co">#&gt; 1 Lomandra longifolia native</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="closer-look">Closer look at name standardisation with ‘APCalign’<a class="anchor" aria-label="anchor" href="#closer-look"></a>
</h2>
<p><code>create_taxonomic_update_lookup</code> is a simple, wrapper,
function for novice users that want to quickly check and standardise
taxon names. For more experienced users, you can take a look at the sub
functions <code>match_taxa()</code>, <code><a href="../reference/align_taxa.html">align_taxa()</a></code> and
<code><a href="../reference/update_taxonomy.html">update_taxonomy()</a></code> to see how taxon names are processed,
aligned and updated.</p>
<p><img src="../reference/figures/standardise_taxonomy_workflow.png"></p>
<div class="section level3">
<h3 id="aligning-names-to-apc-and-apni">Aligning names to APC and APNI<a class="anchor" aria-label="anchor" href="#aligning-names-to-apc-and-apni"></a>
</h3>
<p>The function <code>align_taxa</code> will:</p>
<ol style="list-style-type: decimal">
<li>Clean up your taxonomic names
<ul>
<li>The functions <code>standardise_names</code>,
<code>strip_names</code> and <code>strip_names_extra</code> standardise
infraspecific taxon designations and clean up punctuation and
whitespaces</li>
</ul>
</li>
<li>Find best alignment with APC or APNI to your taxonomic name using
our the function <a href="https://traitecoevo.github.io/APCalign/articles/updating-taxon-names.html">match_taxa</a>
<ul>
<li>A taxonomic name flows through a progression of <a href="https://traitecoevo.github.io/APCalign/articles/updating-taxon-names.html">50
match algorithms</a> until it is able to be aligned to a name on either
the APC or APNI list.<br>
</li>
<li>These include <a href="#fuzzy-match">exact and fuzzy matches</a>.
Fuzzy matches are designed to capture small spelling mistakes and syntax
errors in phrase names.<br>
</li>
<li>These include matches to the entire name string and matches on just
select words in the sequence.<br>
</li>
<li>The sequence of matches has been carefully curated to align names
with the fewest mistakes.</li>
</ul>
</li>
<li>Determine the <code>taxon_rank</code> to which the name can be
resolved, based on its syntax.
<ul>
<li>For names that can only be resolved to genus, reformats the name to
offer a standardised <code>genus sp.</code> name, with additional
information/notes provided as part of the original name in square
brackets, as in <code>Acacia sp. [skinny leaves]</code> or
<code>Acacia sp. [Broken Hill]</code>
</li>
</ul>
</li>
<li>Determine the <code>taxonomic_reference</code> (APC or APNI) of each
name-alignment.</li>
</ol>
<p><strong>Note</strong> that <code>align_taxa</code> <strong>does
not</strong> seek to update outdated taxonomy. That process occurs
during <a href="#update">update_taxonomy</a> process.
<code>align_taxa</code> instead aligns each name input to the closest
match amongst names documented by the APC and APNI.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">aligned_gbif_taxa</span> <span class="op">&lt;-</span> <span class="va">gbif_lite</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">species</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/align_taxa.html">align_taxa</a></span><span class="op">(</span>resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking alignments of 121 taxa</span></span>
<span><span class="co">#&gt;   -&gt; of these 112 names have a perfect match to a scientific name in the APC. </span></span>
<span><span class="co">#&gt;       Alignments being sought for remaining names.</span></span>
<span></span>
<span><span class="va">aligned_gbif_taxa</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 129 × 7</span></span>
<span><span class="co">#&gt;   original_name           cleaned_name            aligned_name    taxonomic_dataset taxon_rank aligned_reason alignment_code</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                   &lt;chr&gt;                   &lt;chr&gt;           &lt;chr&gt;             &lt;chr&gt;      &lt;chr&gt;          &lt;chr&gt;         </span></span>
<span><span class="co">#&gt; 1 Tetratheca ciliata      Tetratheca ciliata      Tetratheca cil… APC               species    Exact match o… match_01c_acc…</span></span>
<span><span class="co">#&gt; 2 Peganum harmala         Peganum harmala         Peganum harmala APC               species    Exact match o… match_01c_acc…</span></span>
<span><span class="co">#&gt; 3 Calotis multicaulis     Calotis multicaulis     Calotis multic… APC               species    Exact match o… match_01c_acc…</span></span>
<span><span class="co">#&gt; 4 Leptospermum trinervium Leptospermum trinervium Leptospermum t… APC               species    Exact match o… match_01c_acc…</span></span>
<span><span class="co">#&gt; 5 Lepidosperma laterale   Lepidosperma laterale   Lepidosperma l… APC               species    Exact match o… match_01c_acc…</span></span>
<span><span class="co">#&gt; 6 Enneapogon polyphyllus  Enneapogon polyphyllus  Enneapogon pol… APC               species    Exact match o… match_01c_acc…</span></span>
<span><span class="co">#&gt; # ℹ 123 more rows</span></span></code></pre></div>
<p>For every <code>aligned_name</code>, <code><a href="../reference/align_taxa.html">align_taxa()</a></code> will
provide a <code>aligned_reason</code> which you can review as a table of
counts:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sfirke/janitor" class="external-link">janitor</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">aligned_gbif_taxa</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">aligned_reason</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://sfirke.github.io/janitor/reference/tabyl.html" class="external-link">tabyl</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="co">#&gt; # A tibble: 6 × 4</span></span>
<span><span class="co">#&gt;   `pull(aligned_gbif_taxa, aligned_reason)`                                                          n percent valid_percent</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                                                                                          &lt;int&gt;   &lt;dbl&gt;         &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 Exact match of taxon name to an APC-accepted canonical name once punctuation and filler words…   118 0.915         0.929  </span></span>
<span><span class="co">#&gt; 2 Exact match of taxon name to an APC-known canonical name once punctuation and filler words ar…     6 0.0465        0.0472 </span></span>
<span><span class="co">#&gt; 3 Exact match of taxon name to an APNI-listed canonical name once punctuation and filler words …     1 0.00775       0.00787</span></span>
<span><span class="co">#&gt; 4 Exact match of the first two words of the taxon name to an APC-accepted canonical name (2024-…     1 0.00775       0.00787</span></span>
<span><span class="co">#&gt; 5 Exact match of the first word of the taxon name to an APC-accepted genus (2024-11-14)              1 0.00775       0.00787</span></span>
<span><span class="co">#&gt; 6 &lt;NA&gt;                                                                                               2 0.0155       NA</span></span></code></pre></div>
<div class="section level4">
<h4 id="fuzzy-match">Configuring matching precision and aligned output<a class="anchor" aria-label="anchor" href="#fuzzy-match"></a>
</h4>
<p>There are arguments in <code>align_taxa</code> that allows you to
select which of the 50 matching algorithms are activated/deactivated and
the degree of fuzziness of the fuzzy matching function</p>
<ul>
<li>
<code>fuzzy_matches</code> turns fuzzy matching on / off (it
defaults to <code>TRUE</code>).<br>
</li>
<li>
<code>fuzzy_abs_dist</code> and <code>fuzzy_rel_dist</code> control
the degree of fuzzy matching (they default to
<code>fuzzy_abs_dist = 3</code> &amp;
<code>fuzzy_rel_dist = 0.2</code>).<br>
</li>
<li>
<code>imprecise_fuzzy_matches</code> turns imprecise fuzzy matching
on / off (it defaults to <code>FALSE</code>; for true it is set to
<code>fuzzy_abs_dist = 5</code> &amp;
<code>fuzzy_rel_dist = 0.25</code>).<br>
</li>
<li>
<code>APNI_matches</code> turns matches to the APNI list on/off (it
defaults to <code>TRUE</code>).<br>
</li>
<li>
<code>identifier</code> allows you to specify a text string that is
added to genus-level matches, indicating the site, study, etc
e.g. <code>Acacia sp. [Blue Mountains]</code>
</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="update">Updating to APC-accepted names<a class="anchor" aria-label="anchor" href="#update"></a>
</h3>
<p><code><a href="../reference/update_taxonomy.html">update_taxonomy()</a></code> uses the information generated by
<code><a href="../reference/align_taxa.html">align_taxa()</a></code> to, whenever possible, update names to
APC-accepted names.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">updated_gbif_taxa</span> <span class="op">&lt;-</span> <span class="va">aligned_gbif_taxa</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/update_taxonomy.html">update_taxonomy</a></span><span class="op">(</span>resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span></span>
<span></span>
<span><span class="va">updated_gbif_taxa</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 129 × 21</span></span>
<span><span class="co">#&gt;   original_name         aligned_name accepted_name suggested_name genus family taxon_rank taxonomic_dataset taxonomic_status</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                 &lt;chr&gt;        &lt;chr&gt;         &lt;chr&gt;          &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;             &lt;chr&gt;           </span></span>
<span><span class="co">#&gt; 1 Tetratheca ciliata    Tetratheca … Tetratheca c… Tetratheca ci… Tetr… Elaeo… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 2 Peganum harmala       Peganum har… Peganum harm… Peganum harma… Pega… Nitra… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 3 Calotis multicaulis   Calotis mul… Calotis mult… Calotis multi… Calo… Aster… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 4 Leptospermum trinerv… Leptospermu… Leptospermum… Leptospermum … Lept… Myrta… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 5 Lepidosperma laterale Lepidosperm… Lepidosperma… Lepidosperma … Lepi… Cyper… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 6 Enneapogon polyphyll… Enneapogon … Enneapogon p… Enneapogon po… Enne… Poace… species    APC               accepted        </span></span>
<span><span class="co">#&gt; # ℹ 123 more rows</span></span>
<span><span class="co">#&gt; # ℹ 12 more variables: taxonomic_status_aligned &lt;chr&gt;, aligned_reason &lt;chr&gt;, update_reason &lt;chr&gt;, subclass &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   taxon_distribution &lt;chr&gt;, scientific_name &lt;chr&gt;, taxon_ID &lt;chr&gt;, taxon_ID_genus &lt;chr&gt;, scientific_name_ID &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   canonical_name &lt;chr&gt;, row_number &lt;dbl&gt;, number_of_collapsed_taxa &lt;dbl&gt;</span></span></code></pre></div>
<div class="section level4">
<h4 id="taxonomic-resources-used-for-updating-names">Taxonomic resources used for updating names<a class="anchor" aria-label="anchor" href="#taxonomic-resources-used-for-updating-names"></a>
</h4>
<ul>
<li><p>The APC includes all previously recorded taxonomic names for a
current taxon concept, designating the currently-accepted name as
<code>taxonomic_status: accepted</code>, while previously used or
inappropriately used names for the taxon concept have alternative
taxonomic statuses documented (e.g. taxonomic synonym, orthographic
variant, misapplied).</p></li>
<li><p>The APC includes a column <code>acceptedNameUsageID</code> that
links a taxon name with an alternative taxonomic status to the current
taxon name, allowing outdated/inappropriately used names to be synced to
their current name.</p></li>
</ul>
<p><em>Note</em>: Names listed on the APNI but absent from the APC are
those that are designated as <code>taxonomic_dataset: APNI</code> by
<code>APCalign</code>. These are names that are currently
<code>unknown</code> by the APC. Over time, this list shrinks, as
taxonomists link ever more occasionally used name variants to an
APC-accepted taxon. However, for now, <em>names listed only on the APNI
cannot be updated</em></p>
</div>
<div class="section level4">
<h4 id="name-updates-at-different-taxonomic-levels">Name updates at different taxonomic levels<a class="anchor" aria-label="anchor" href="#name-updates-at-different-taxonomic-levels"></a>
</h4>
<ul>
<li>
<code><a href="../reference/update_taxonomy.html">update_taxonomy()</a></code> divides names into lists based on the
<code>taxon_rank</code> and <code>taxonomic_dataset</code> assigned by
<code>align_taxa</code>, as each list requires different updating
algorithms.<br>
</li>
<li>Only taxonomic names that are designated as
<code>taxon_rank = species/infraspecific</code> and
<code>taxonomic_dataset = APC</code> can be updated to an APC-accepted
name.<br>
</li>
<li>For all other taxa, it may be possible to align the genus-name to an
APC-accepted genus.<br>
</li>
<li>For all taxa, a <code>suggested_name</code> is provided, selecting
the <code>accepted_name</code> when available, and otherwise the
<code>aligned_name</code>, but with, if possible, an updated,
APC-accepted genus name.</li>
</ul>
</div>
<div class="section level4">
<h4 id="taxonomic-splits">Taxonomic splits<a class="anchor" aria-label="anchor" href="#taxonomic-splits"></a>
</h4>
<ul>
<li><p>Taxonomic splits refers to instances where a single taxon concept
is subsequently split into multiple taxon concepts. For such taxa, when
the <code>aligned_name</code> is the “old” taxon concept name, it is
impossible to know which of the currently accepted taxon concepts the
name represents.</p></li>
<li>
<p>The function <code>update_taxonomy</code> includes an argument
<code>taxonomic_splits</code>, offering three alternative outputs for
taxon concepts that have been split.</p>
<ol style="list-style-type: decimal">
<li><p><code>most_likely_species</code> is the default value, and
returns the <code>accepted_name</code> of the original taxon_concept;
alternative names are documented in square brackets as part of the
suggested name
(<code>Acacia aneura [alternative possible names: Acacia minyura (pro parte misapplied) | Acacia paraneura (pro parte misapplied) | Acacia quadrimarginea (misapplied)</code>).</p></li>
<li><p><code>return_all</code> returns all currently accepted names that
were split from the original taxon_concept; this leads to an increase in
the number of rows in the output table. (Acacia aneura, Acacia minyura
and Acacia paraneura are each output as a separate row, each with a
unique taxon_ID)</p></li>
<li><p><code>collapse_to_higher_taxon</code> declares that for split
names, there is no way to be certain about which accepted name is
appropriate and therefore that the best possible match is at the genus
level; no <code>accepted_name</code> is returned, the
<code>taxon_rank</code> is demoted to <code>genus</code> and the
suggested name documents the possible species-level names in square
brackets
(<code>Acacia sp. [collapsed names: Acacia aneura (accepted) | Acacia minyura (pro parte misapplied) | Acacia paraneura (pro parte misapplied)]</code>)</p></li>
</ol>
</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">aligned_gbif_taxa</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/update_taxonomy.html">update_taxonomy</a></span><span class="op">(</span>taxonomic_splits <span class="op">=</span> <span class="st">"most_likely_species"</span>, </span>
<span>                  resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>  <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">original_name</span> <span class="op">==</span> <span class="st">"Acacia aneura"</span><span class="op">)</span>  <span class="co"># Subsetting Acacia aneura as an example </span></span>
<span><span class="co">#&gt; # A tibble: 1 × 21</span></span>
<span><span class="co">#&gt;   original_name aligned_name  accepted_name suggested_name        genus family taxon_rank taxonomic_dataset taxonomic_status</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;         &lt;chr&gt;         &lt;chr&gt;                 &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;             &lt;chr&gt;           </span></span>
<span><span class="co">#&gt; 1 Acacia aneura Acacia aneura Acacia aneura Acacia aneura [alter… Acac… Fabac… species    APC               accepted        </span></span>
<span><span class="co">#&gt; # ℹ 12 more variables: taxonomic_status_aligned &lt;chr&gt;, aligned_reason &lt;chr&gt;, update_reason &lt;chr&gt;, subclass &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   taxon_distribution &lt;chr&gt;, scientific_name &lt;chr&gt;, taxon_ID &lt;chr&gt;, taxon_ID_genus &lt;chr&gt;, scientific_name_ID &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   canonical_name &lt;chr&gt;, row_number &lt;dbl&gt;, number_of_collapsed_taxa &lt;dbl&gt;</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aligned_gbif_taxa</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/update_taxonomy.html">update_taxonomy</a></span><span class="op">(</span>taxonomic_splits <span class="op">=</span> <span class="st">"return_all"</span>,</span>
<span>                  resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>  <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">original_name</span> <span class="op">==</span> <span class="st">"Acacia aneura"</span><span class="op">)</span> <span class="co"># Subsetting Acacia aneura as an example</span></span>
<span><span class="co">#&gt; # A tibble: 3 × 21</span></span>
<span><span class="co">#&gt;   original_name aligned_name  accepted_name    suggested_name   genus  family  taxon_rank taxonomic_dataset taxonomic_status</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;         &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;             &lt;chr&gt;           </span></span>
<span><span class="co">#&gt; 1 Acacia aneura Acacia aneura Acacia aneura    Acacia aneura    Acacia Fabace… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 2 Acacia aneura Acacia aneura Acacia minyura   Acacia minyura   Acacia Fabace… species    APC               accepted        </span></span>
<span><span class="co">#&gt; 3 Acacia aneura Acacia aneura Acacia paraneura Acacia paraneura Acacia Fabace… species    APC               accepted        </span></span>
<span><span class="co">#&gt; # ℹ 12 more variables: taxonomic_status_aligned &lt;chr&gt;, aligned_reason &lt;chr&gt;, update_reason &lt;chr&gt;, subclass &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   taxon_distribution &lt;chr&gt;, scientific_name &lt;chr&gt;, taxon_ID &lt;chr&gt;, taxon_ID_genus &lt;chr&gt;, scientific_name_ID &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   canonical_name &lt;chr&gt;, row_number &lt;dbl&gt;, number_of_collapsed_taxa &lt;dbl&gt;</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aligned_gbif_taxa</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/update_taxonomy.html">update_taxonomy</a></span><span class="op">(</span>taxonomic_splits <span class="op">=</span> <span class="st">"collapse_to_higher_taxon"</span>,</span>
<span>                  resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>  <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">original_name</span> <span class="op">==</span> <span class="st">"Acacia aneura"</span><span class="op">)</span> <span class="co"># Subsetting Acacia aneura as an example</span></span>
<span><span class="co">#&gt; # A tibble: 1 × 21</span></span>
<span><span class="co">#&gt;   original_name aligned_name  accepted_name suggested_name        genus family taxon_rank taxonomic_dataset taxonomic_status</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;         &lt;chr&gt;         &lt;chr&gt;                 &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;             &lt;chr&gt;           </span></span>
<span><span class="co">#&gt; 1 Acacia aneura Acacia aneura Acacia sp.    Acacia sp. [collapse… Acac… Fabac… species    APC               accepted        </span></span>
<span><span class="co">#&gt; # ℹ 12 more variables: taxonomic_status_aligned &lt;chr&gt;, aligned_reason &lt;chr&gt;, update_reason &lt;chr&gt;, subclass &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   taxon_distribution &lt;chr&gt;, scientific_name &lt;chr&gt;, taxon_ID &lt;chr&gt;, taxon_ID_genus &lt;chr&gt;, scientific_name_ID &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   canonical_name &lt;chr&gt;, row_number &lt;dbl&gt;, number_of_collapsed_taxa &lt;dbl&gt;</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel Falster, Elizabeth Wenk, Will Cornwell, Fonti Kar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
