<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="APCalign">
<title>Function notes • APCalign</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function notes">
<meta property="og:description" content="APCalign">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">APCalign</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/APCalign.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data providers</h6>
    <a class="dropdown-item" href="../articles/data-providers.html">APC and APNI</a>
    <h6 class="dropdown-header" data-toc-skip>Functions</h6>
    <a class="dropdown-item" href="../articles/function_notes.html">Details on the 10 exported functions, including examples of usage</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Taxon matching</h6>
    <a class="dropdown-item" href="../articles/updating-taxon-names.html">Our fuzzy matching  algorithm</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/traitecoevo/APCalign/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function notes</h1>
                        <h4 data-toc-skip class="author">Elizabeth
Wenk</h4>
            
            <h4 data-toc-skip class="date">2024-01-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/traitecoevo/APCalign/blob/HEAD/vignettes/articles/function_notes.Rmd" class="external-link"><code>vignettes/articles/function_notes.Rmd</code></a></small>
      <div class="d-none name"><code>function_notes.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="apcalign-functions">APCalign functions<a class="anchor" aria-label="anchor" href="#apcalign-functions"></a>
</h2>
<p>APCalign exports <a href="https://traitecoevo.github.io/APCalign/reference/index.html">10
functions</a> to facilitate the alignment of submitted plant names to
scientific names on the APC and APNI lists. They are listed in order of
likelihood of use.</p>
<div class="section level3">
<h3 id="taxon-name-alignment-and-updating-functions">Taxon name alignment and updating functions<a class="anchor" aria-label="anchor" href="#taxon-name-alignment-and-updating-functions"></a>
</h3>
<div class="section level4">
<h4 id="create_taxonomic_update_lookup">create_taxonomic_update_lookup<a class="anchor" aria-label="anchor" href="#create_taxonomic_update_lookup"></a>
</h4>
<p><strong>description</strong>: This function takes a list of
Australian plant names that need to be reconciled with current taxonomy
and generates a lookup table of the best-possible scientific name match
for each input name. It uses first the function <code>align_taxa</code>,
then the function <code>update_taxonomy</code> to achieve the output.
The aligned name is plant name that has been aligned to a taxon name in
the APC or APNI by the align_taxa function.</p>
<p><strong>usage notes</strong>: This is APCalign’s core function,
merging together the alignment and updating of taxonomy.</p>
<p><strong>arguments</strong>:</p>
<pre><code><span><span class="va">taxa</span>                                      <span class="co">#input vector of taxon names</span></span>
<span><span class="va">stable_or_current_data</span> <span class="op">=</span> <span class="st">"stable"</span>  </span>
<span><span class="va">version</span> <span class="op">=</span> <span class="fu">default_version</span><span class="op">(</span><span class="op">)</span>  </span>
<span><span class="va">taxonomic_splits</span> <span class="op">=</span> <span class="st">"most_likely_species"</span>  <span class="co">#options for names with ambiguous taxonomic histories</span></span>
<span><span class="va">full</span> <span class="op">=</span> <span class="cn">FALSE</span>                              <span class="co">#outputs fewer (FALSE) or more (TRUE) columns</span></span>
<span><span class="va">APNI_matches</span> <span class="op">=</span> <span class="cn">TRUE</span>                       <span class="co">#include (TRUE) or exclude (FALSE) APNI list</span></span>
<span><span class="va">imprecise_fuzzy_matches</span> <span class="op">=</span> <span class="cn">FALSE</span>           <span class="co">#disallow (FALSE) or allow (TRUE) imprecise fuzzy matches</span></span>
<span><span class="va">identifier</span> <span class="op">=</span> <span class="cn">NA_character_</span>                <span class="co">#include a unique identifier as part of informal names</span></span>
<span><span class="va">resources</span> <span class="op">=</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span><span class="op">)</span>  </span>
<span><span class="va">output</span> <span class="op">=</span> <span class="cn">NULL</span>  </span></code></pre>
<p><strong>output</strong>: A data frame with rows representing each
taxon and columns documenting taxon metadata (<em>original_name,
aligned_name, accepted_name, suggested_name, genus, family, taxon_rank,
taxonomic_dataset, taxonomic_status, taxonomic_status_aligned,
aligned_reason, update_reason, subclass, taxon_distribution,
scientific_name_authorship, taxon_ID, taxon_ID_genus,
scientific_name_ID, row_number, number_of_collapsed_taxa</em>).</p>
<p><strong>example</strong>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Banksia serrata"</span>, <span class="st">"Banksia serrate"</span>, <span class="st">"Banksia cerrata"</span>, <span class="st">"Banksea serrata"</span>, <span class="st">"Banksia serrrrata"</span>, <span class="st">"Dryandra"</span><span class="op">)</span></span>
<span><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">updated_taxa</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">APCalign</span><span class="fu">::</span><span class="fu"><a href="../reference/create_taxonomic_update_lookup.html">create_taxonomic_update_lookup</a></span><span class="op">(</span></span>
<span>    taxa <span class="op">=</span> <span class="va">input</span>,</span>
<span>    identifier <span class="op">=</span> <span class="st">"APCalign test"</span>,</span>
<span>    full <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    resources <span class="op">=</span> <span class="va">resources</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>or, start with a csv file where there is a column of taxon names to
align</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxon_list</span> <span class="op">&lt;-</span>                          <span class="co">#or load data through the R studio menu</span></span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu">here</span><span class="op">(</span><span class="st">"inst/"</span>, <span class="st">"extdata"</span>, <span class="st">"test_taxa.csv"</span><span class="op">)</span>,</span>
<span>    show_col_types <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">updated_taxa</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">APCalign</span><span class="fu">::</span><span class="fu"><a href="../reference/create_taxonomic_update_lookup.html">create_taxonomic_update_lookup</a></span><span class="op">(</span></span>
<span>    taxa <span class="op">=</span> <span class="va">taxon_list</span><span class="op">$</span><span class="va">original_name</span>,</span>
<span>    identifier <span class="op">=</span> <span class="st">"APCalign test"</span>,</span>
<span>    full <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    resources <span class="op">=</span> <span class="va">resources</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><strong>notes</strong><br>
- If you will be running the function
<code><a href="../reference/create_taxonomic_update_lookup.html">APCalign::create_taxonomic_update_lookup</a></code> many times, it is
best to load the taxonomic resources separately using
<code>resources &lt;- load_taxonomic_resources()</code>, then add the
argument <code>resources = resources</code><br>
- The name <code>Banksia cerrata</code> does not align as the fuzzy
matching algorithm does not allow the first letter of the genus and
species epithet to change.<br>
- The argument <code>taxonomic_splits</code> allows you to choose the
outcome for updating the names of taxa with ambiguous taxonomic
histories; this applies to scientific names that were once attached to a
more broadly circumscribed taxon concept, that was then split into
several more narrowly circumscribed taxon concepts, one of which retains
the original name. There are three options:
<code>most_likely_species</code> returns the name that is retained, with
alternative names documented in square brackets; <code>return_all</code>
adds additional rows to the output, one for each possible taxon concept;
<code>collapse_to_higher_taxon</code> returns the genus with possible
names in square brackets.<br>
- The argument <code>identifier</code> allows you to add a fix text
string to all genus- and family- level names, such as
<code>identifier = "Royal NP"</code> would return `Acacia sp. [Royal
NP]`.</p>
</div>
<div class="section level4">
<h4 id="align_taxa">align_taxa<a class="anchor" aria-label="anchor" href="#align_taxa"></a>
</h4>
<p><strong>description</strong>: This function finds taxonomic
alignments in the APC or APNI. It uses the internal function
<code>match_taxa</code> to attempt to match input strings to taxon names
in the APC/APNI. It sequentially searches for matches against more than
20 different string patterns, prioritising exact matches (to accepted
names as well as synonyms, orthographic variants) over fuzzy matches. It
prioritises matches to taxa in the APC over names in the APNI. It
identifies string patterns in input names that suggest a name can only
be aligned to a genus (hybrids that are not in the APC/ANI; graded
species; taxa not identified to species), and indicates these names only
have a genus-rank match.</p>
<p><strong>usage notes</strong>: Users will run this function if they
wish to see the details of the matching algorithms, the many output
columns that the matching function compares to as it seeks the best
alignment. They may also select this function if they want to adjust the
“fuzziness” level for fuzzy matches, options not allowed in
<code>create_taxonomic_update_lookup</code>. This function is the first
half of <code>create_taxonomic_update_lookup</code>.</p>
<p><strong>arguments</strong>:</p>
<pre><code><span><span class="va">original_name</span>                           <span class="co">#input vector of taxon names</span></span>
<span><span class="va">output</span> <span class="op">=</span> <span class="cn">NULL</span>  </span>
<span><span class="va">full</span> <span class="op">=</span> <span class="cn">FALSE</span>                            <span class="co">#outputs fewer (FALSE) or more (TRUE) columns</span></span>
<span><span class="va">resources</span> <span class="op">=</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span><span class="op">)</span>  </span>
<span><span class="va">fuzzy_abs_dist</span> <span class="op">=</span> <span class="fl">3</span>                      <span class="co">#set number of characters allowed to be different for fuzzy match</span></span>
<span><span class="va">fuzzy_rel_dist</span> <span class="op">=</span> <span class="fl">0.2</span>                    <span class="co">#set proportion of characters allowed to be different for fuzzy match</span></span>
<span><span class="va">fuzzy_matches</span> <span class="op">=</span> <span class="cn">TRUE</span>                    <span class="co">#disallow (FALSE) or allow (TRUE) any fuzzy matches</span></span>
<span><span class="va">imprecise_fuzzy_matches</span> <span class="op">=</span> <span class="cn">FALSE</span>         <span class="co">#disallow (FALSE) or allow (TRUE) imprecise fuzzy matches</span></span>
<span><span class="va">APNI_matches</span> <span class="op">=</span> <span class="cn">TRUE</span>                     <span class="co">#include (TRUE) or exclude (FALSE) APNI list</span></span>
<span><span class="va">identifier</span> <span class="op">=</span> <span class="va">NA_character</span>               <span class="co">#include a unique identifier as part of informal names</span></span></code></pre>
<p><strong>output</strong>: A data frame with rows representing each
taxon and with columns documenting the alignment made, the reason for
this alignment, and a selection of taxon name mutations to which the
original name was compared (<em>original_name, aligned_name,
taxonomic_dataset, taxon_rank, aligned_reason, alignment_code,
cleaned_name, stripped_name, stripped_name2, trinomial, binomial, genus,
fuzzy_match_genus, fuzzy_match_genus_synonym, fuzzy_match_genus_APNI,
fuzzy_match_cleaned_APC, fuzzy_match_cleaned_APC_synonym,
fuzzy_match_cleaned_APC_imprecise,
fuzzy_match_cleaned_APC_synonym_imprecise, fuzzy_match_binomial,
fuzzy_match_binomial_APC_synonym, fuzzy_match_trinomial,
fuzzy_match_trinomial_synonym, fuzzy_match_cleaned_APNI,
fuzzy_match_cleaned_APNI_imprecise</em>).</p>
<p><strong>example</strong>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Banksia serrata"</span>, <span class="st">"Banksia serrate"</span>, <span class="st">"Banksia cerrata"</span>, <span class="st">"Banksia serrrrata"</span>, <span class="st">"Dryandra sp."</span>, <span class="st">"Banksia big red flowers"</span><span class="op">)</span></span>
<span><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">aligned_taxa</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">APCalign</span><span class="fu">::</span><span class="fu"><a href="../reference/align_taxa.html">align_taxa</a></span><span class="op">(</span></span>
<span>    original_name <span class="op">=</span> <span class="va">input</span>,</span>
<span>    identifier <span class="op">=</span> <span class="st">"APCalign test"</span>,</span>
<span>    full <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    resources <span class="op">=</span> <span class="va">resources</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><strong>notes</strong><br>
- If you will be running the function
<code><a href="../reference/create_taxonomic_update_lookup.html">APCalign::create_taxonomic_update_lookup</a></code> many times, it is
best to load the taxonomic resources separately using
<code>resources &lt;- load_taxonomic_resources()</code>, then add the
argument <code>resources = resources</code><br>
- The name <code>Banksia cerrata</code> does not align as the fuzzy
matching algorithm does not allow the first letter of the genus and
species epithet to change.<br>
- With this function you have the option of changing the fuzzy matching
parameters. The defaults, with fuzzy matches only allowing changes of 3
(or fewer) characters AND 20% (or less) of characters has been carefully
calibrated to catch just about all typos, but very, very rarely
mis-align a name. If you wish to introduce less conservative fuzzy
matching it is recommended you manually check the aligned names.<br>
- It is recommended that you begin with
<code>imprecise_fuzzy_matches = FALSE</code> (the default), as quite a
few of the less precise fuzzy matches are likely to be erroneous. This
argument should be turned on only if you plan to check all alignments
manually.<br>
- The argument <code>identifier</code> allows you to add a fix text
string to all genus- and family- level names, such as
<code>identifier = "Royal NP"</code> would return
<code>Acacia sp. [Royal NP]</code>.</p>
</div>
<div class="section level4">
<h4 id="update_taxonomy">update_taxonomy<a class="anchor" aria-label="anchor" href="#update_taxonomy"></a>
</h4>
<p><strong>description</strong>: This function uses the APC to update
the taxonomy of names aligned to a taxon concept listed in the APC to
the currently accepted name for the taxon concept. The aligned_data data
frame that is input must contain 5 columns, <code>originial_name</code>,
<code>aligned_name</code>, <code>taxon_rank</code>,
<code>taxonomic_dataset</code>, and <code>aligned_reason</code>, the
columns output by the function <code><a href="../reference/align_taxa.html">APCalign::align_taxa()</a></code>. The
aligned name is a plant name that has been aligned to a taxon name in
the APC or APNI by the align_taxa function.</p>
<p><strong>usage notes</strong>: As the input for this function is a
table with 5 columns (output by <code>align_taxa</code>), this function
will only be used when you explicitly want to separate the
<code>aligment</code> and <code>updating</code> components of APCalign.
This function is the second half of
<code>create_taxonomic_update_lookup</code>.</p>
<p><strong>arguments</strong>:</p>
<pre><code><span><span class="va">aligned_data</span>                              <span class="co">#input table of aligned names and information about the aligned name</span></span>
<span><span class="va">taxonomic_splits</span> <span class="op">=</span> <span class="st">"most_likely_species"</span>  <span class="co">#options for names with ambiguous taxonomic histories</span></span>
<span><span class="va">output</span> <span class="op">=</span> <span class="cn">NULL</span>   </span>
<span><span class="va">resources</span> <span class="op">=</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span><span class="op">)</span>  </span></code></pre>
<p><strong>output</strong>: A data frame with rows representing each
taxon and columns documenting taxon metadata (<em>original_name,
aligned_name, accepted_name, suggested_name, genus, family, taxon_rank,
taxonomic_dataset, taxonomic_status, taxonomic_status_aligned,
aligned_reason, update_reason, subclass, taxon_distribution,
scientific_name_authorship, taxon_ID, taxon_ID_genus,
scientific_name_ID, row_number, number_of_collapsed_taxa</em>).</p>
</div>
</div>
<div class="section level3">
<h3 id="diversity-and-distribution-functions">Diversity and distribution functions<a class="anchor" aria-label="anchor" href="#diversity-and-distribution-functions"></a>
</h3>
<div class="section level4">
<h4 id="create_species_state_origin_matrix">create_species_state_origin_matrix<a class="anchor" aria-label="anchor" href="#create_species_state_origin_matrix"></a>
</h4>
<p><strong>description</strong>: This function processes the geographic
data available in the APC and returns state level native, introduced and
more complicated origins status for all taxa.</p>
<p><strong>arguments</strong>:</p>
<pre><code><span><span class="va">resources</span> <span class="op">=</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span><span class="op">)</span>  </span></code></pre>
<p><strong>output</strong>: A data frame with rows representing each
species and columns for taxon name and each state . The values in each
cell represent the origin of the species in that state.</p>
</div>
<div class="section level4">
<h4 id="native_anywhere_in_australia">native_anywhere_in_australia<a class="anchor" aria-label="anchor" href="#native_anywhere_in_australia"></a>
</h4>
<p><strong>description</strong>: This function checks if the given
species is native anywhere in Australia according to the APC. Note that
this will not detect within-Australia introductions, e.g. if a species
is from Western Australia and is invasive on the east coast.</p>
<p><strong>arguments</strong>:</p>
<pre><code><span><span class="va">species</span>                                  <span class="co">#input vector of taxon names</span></span>
<span><span class="va">resources</span> <span class="op">=</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span><span class="op">)</span>  </span></code></pre>
<p><strong>output</strong>: A data frame with rows representing each
taxon and two columns: <code>species</code>, which is the same as the
unique values of the input <code>species</code>, and
<code>native_anywhere_in_aus</code>, a vector indicating whether each
species is native anywhere in Australia, introduced by humans from
elsewhere, or unknown with respect to the APC resource.</p>
</div>
<div class="section level4">
<h4 id="state_diversity_counts">state_diversity_counts<a class="anchor" aria-label="anchor" href="#state_diversity_counts"></a>
</h4>
<p><strong>description</strong>: This function calculates state-level
diversity for native, introduced, and more complicated species origins
based on the geographic data available in the APC.</p>
<p><strong>arguments</strong>:</p>
<pre><code><span><span class="va">state</span>                                    <span class="co">#state for which diversity should be summarised</span></span>
<span><span class="va">resources</span> <span class="op">=</span> <span class="fu"><a href="../reference/load_taxonomic_resources.html">load_taxonomic_resources</a></span><span class="op">(</span><span class="op">)</span>  </span></code></pre>
<p><strong>output</strong>: A data frame with three columns: “origin”
indicating the origin of the species, “state” indicating the Australian
state or territory, and “num_species” indicating the number of species
for that origin and state.</p>
</div>
</div>
<div class="section level3">
<h3 id="utility-functions">Utility functions<a class="anchor" aria-label="anchor" href="#utility-functions"></a>
</h3>
<div class="section level4">
<h4 id="load_taxonomic_resources">load_taxonomic_resources<a class="anchor" aria-label="anchor" href="#load_taxonomic_resources"></a>
</h4>
<p><strong>description</strong>: This function loads two taxonomic
datasets for Australia’s vascular plants, the APC and APNI, into the
global environment. It accesses taxonomic data from a dataset using the
provided version number or the default version. The function creates
several data frames by filtering and selecting data from the loaded
lists.</p>
<p><strong>usage notes</strong>: This function is called by many other
APC functions, but is unlikely to be used independently by a APCalign
user.</p>
<p><strong>arguments</strong>:</p>
<pre><code><span><span class="va">stable_or_current_data</span> <span class="op">=</span> <span class="st">"stable"</span>  </span>
<span><span class="va">version</span> <span class="op">=</span> <span class="fu">default_version</span><span class="op">(</span><span class="op">)</span>  </span>
<span><span class="va">reload</span> <span class="op">=</span> <span class="cn">FALSE</span>  </span></code></pre>
<p><strong>output</strong>: Several dataframes that include subsets of
the APC/APNI based on taxon rank and taxonomic status.</p>
</div>
<div class="section level4">
<h4 id="standardise_names">standardise_names<a class="anchor" aria-label="anchor" href="#standardise_names"></a>
</h4>
<p><strong>description</strong>: This function standardises taxon names
by performing a series of text substitutions to remove common
inconsistencies in taxonomic nomenclature. The function takes a
character vector of taxon names as input and returns a character vector
of taxon names using standardised taxonomic syntax as output. In
particular it standardises taxon rank abbreviations and qualifiers
(subsp., var., f.), as people use many variants of these terms. It also
standardises or removes a few additional filler words used within taxon
names (affinis becomes aff.; s.l. and s.s. are removed).</p>
<p><strong>arguments</strong>:</p>
<pre><code><span><span class="va">taxon_names</span>           <span class="co">#input vector of taxon names</span></span></code></pre>
<p><strong>output</strong>: A character vector of standardised taxon
names.</p>
</div>
<div class="section level4">
<h4 id="strip_names">strip_names<a class="anchor" aria-label="anchor" href="#strip_names"></a>
</h4>
<p><strong>description</strong>: Given a vector of taxonomic names, this
function removes subtaxa designations (“subsp.”, “var.”, “f.”, and
“ser”), special characters (e.g., “-”, “.”, “(”, “)”, “?”), and extra
whitespace. The resulting vector of names is also converted to
lowercase.</p>
<p><strong>arguments</strong>:</p>
<pre><code><span><span class="va">taxon_names</span>            <span class="co">#input vector of taxon names   </span></span></code></pre>
<p><strong>output</strong>: A character vector of stripped taxonomic
names, with subtaxa designations, special characters, and extra
whitespace removed, and all letters converted to lowercase.</p>
</div>
<div class="section level4">
<h4 id="strip_names_2">strip_names_2<a class="anchor" aria-label="anchor" href="#strip_names_2"></a>
</h4>
<p><strong>description</strong>: Given a vector of taxonomic names, this
function removes subtaxa designations (“subsp.”, “var.”, “f.”, and
“ser”), additional filler words and characters (” x ” [hybrid taxa],
“sp.”), special characters (e.g., “-”, “.”, “(”, “)”, “?”), and extra
whitespace. The resulting vector of names is also converted to
lowercase.</p>
<p><strong>arguments</strong>:</p>
<pre><code><span><span class="va">taxon_names</span>             <span class="co">#input vector of taxon names   </span></span></code></pre>
<p><strong>output</strong>: A character vector of stripped taxonomic
names, with subtaxa designations, special characters, additional filler
words and extra whitespace removed, and all letters converted to
lowercase.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Daniel Falster, Elizabeth Wenk, Will Cornwell, Fonti Kar.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.8.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
